datafusion/core/src/physical_plan/joins/hash_join.rs

```diff
@@ -1482,105 +1325,115 @@ impl HashJoinStream {
                         left_data,
                         &self.on_left,
                         &self.on_right,
                         &self.filter,
-                        self.join_type,
-                        &self.schema,
-                        &self.column_indices,
                         &self.random_state,
                         &self.null_equals_null,
                     );
-                    self.join_metrics.input_batches.add(1);
-                    self.join_metrics.input_rows.add(batch.num_rows());
-                    if let Ok((ref batch, ref left_side)) = result {
-                        timer.done();
-                        self.join_metrics.output_batches.add(1);
-                        self.join_metrics.output_rows.add(batch.num_rows());
-
-                        match self.join_type {
-                            JoinType::Left
-                            | JoinType::Full
-                            | JoinType::LeftSemi
-                            | JoinType::LeftAnti => {
+
+                    match left_right_indices {
+                        Ok((left_side, right_side)) => {
+                            timer.done();
+
+                            // set the left bitmap
+                            // and only left, full, left semi, left anti need the left bitmap
+                            if need_produce_result_in_final(self.join_type) {
                                 left_side.iter().flatten().for_each(|x| {
                                     visited_left_side.set_bit(x as usize, true);
                                 });
                             }
-                            JoinType::Inner
-                            | JoinType::Right
-                            | JoinType::RightSemi
-                            | JoinType::RightAnti => {}
+
+                            // adjust the two side indices base on the join type
+                            let (left_side, right_side) = adjust_indices_by_join_type(
+                                left_side,
+                                right_side,
+                                batch.num_rows(),
+                                self.join_type,
+                            );
+
+                            let result = build_batch_from_indices(
+                                &self.schema,
+                                &left_data.1,
+                                &batch,
+                                left_side,
+                                right_side,
+                                &self.column_indices,
+                            );
+                            self.join_metrics.output_batches.add(1);
+                            self.join_metrics.output_rows.add(batch.num_rows());
+                            Some(result)
+                        }
+                        Err(_) => {
+                            // TODO why the type of result stream is `Result<T, ArrowError>`, and not the `DataFusionError`
```

Comment:
@alamb  @Dandandan  why we use the `ArrowError` instead the `DataFusionError`

Reply:
Is it same with #4172?

Reply:
> Is it same with #4172?

yes